### PR TITLE
DOC: Groupby transform should mention that parameter can be a string

### PR DESCRIPTION
**Pull request by seanjedi**
- [ ] closes https://github.com/pandas-dev/pandas/issues/49961
- [ ] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [ ] Added [type annotations](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#type-hints) to new arguments/methods/functions.
- [ ] Added an entry in the latest `pandas/core/groupby/groupby.py`

### CODE REVIEW

**Comment by rhshadrach** [comment]
@seanjedi - thanks for the PR! Can you shorten the title (there is an edit button above) - being concise is best here.

**Comment by ramvikrams** [comment]
Can you add examples also

**Comment by ramvikrams** [comment]
For example the example givenin dataframegroupby is
`
grouped.transform("sum")
`
so can you add this with it's output and similarly for SeriesGroupBy.transform()

File: pandas/core/groupby/groupby.py [hunk 1]
```diff
@@ -403,7 +403,10 @@ class providing the base-class of operations.
 f : function, str
     Function to apply to each group. See the Notes section below for requirements.
 
-    Can also accept a Numba JIT function with
+    Accepted inputs are:
+
+    - A String
+    - Numba JIT function with
     ``engine='numba'`` specified.
```
Comment (phofl) [inline pandas/core/groupby/groupby.py:410]: I think you have to add a blank line here

Comment (seanjedi) [reply pandas/core/groupby/groupby.py]: I think that this line needed to be a part of the line above and not on its own separate line, so I moved it up. If that is incorrect, then I apologize and I will make that change. This has been changed here: 3b84281aea5673b2e3ba4624e2b7532895c2860a

**Comment by seanjedi** [comment]
Responding to: https://github.com/pandas-dev/pandas/pull/50029#issuecomment-1336017459 
I believe that the docs for dataframegroupby and SeriesGroupBy transform utilizes the same location for their docs, so it only requires the changes I made in the code. Would you want me to add what to expect for the output for this function in general? 

Also, I have updated the code here for the example with `grouped.transform("sum")`
3b84281aea5673b2e3ba4624e2b7532895c2860a

**Comment by ramvikrams** [comment]
> @ramvikrams  Responding to: https://github.com/pandas-dev/pandas/pull/50029#issuecomment-1336017459 
> I believe that the docs for dataframegroupby and SeriesGroupBy transform utilizes the same location for their docs in the code, so it only requires the changes I made in that location to show up for both, as this is a template that get's used by those respective classes.
> Would you want me to add what to expect for the output for this function in general? (I think it already does that?)
>  Or for dataframegroupby  and SeriesGroupBy in particular?
> 
> Also, I have updated the code here for the example with `grouped.transform("sum")`
> 3b84281aea5673b2e3ba4624e2b7532895c2860a

We have the example also so you can add a pd.Series for series and can you also add the output of  `grouped.transform("sum")

**Comment by seanjedi** [comment]
@ramvikrams is this better? 
b06287ea491939ead79f91ea881df8d141b4fd8e

**Comment by ramvikrams** [comment]
> @ramvikrams is this better? 
> b06287ea491939ead79f91ea881df8d141b4fd8e

I think changing the example for series and then rewriting it completely because right now it's with pd.dataframe

**Comment by seanjedi** [comment]
@ramvikrams Sorry, I had to update my comment, for https://github.com/pandas-dev/pandas/commit/b06287ea491939ead79f91ea881df8d141b4fd8e it should be updating the template, so that:
In DataFrameGroupBy.transform it should be `pd.Dataframe = grouped.transform("sum")`  
and in SeriesGroupBy.transform it should be `pd.Series = grouped.transform("sum")`

Since `%(klass)s` seems to be updated in the docs to what is using it, so in DataFrameGroupBy it gets updated to Dataframe, and SeriesGroupBy it gets changed to Series

So I am a bit confused since that is what you wanted correct? 
Or am I mistaking the request? 
Care to elaborate a little bit more, as I am a bit confused by the ask sorry.

**Comment by ramvikrams** [comment]
I am trying to say that this example 
```
df = pd.DataFrame({'A' : ['foo', 'bar', 'foo', 'bar',
                          'foo', 'bar'],
                   'B' : ['one', 'one', 'two', 'three',
                          'two', 'two'],
                   'C' : [1, 5, 5, 2, 5, 5],
                   'D' : [2.0, 5., 8., 1., 2., 9.]})
```
which is there in SeriesGroupBy.transform can you also change this it should be with pd.Series not dataframe.

**Comment by ramvikrams** [comment]
@seanjedi can you change this for the series example
```
ser = pd.Series(
        [390.0, 350.0, 30.0, 20.0],
        index=["Falcon", "Falcon", "Parrot", "Parrot"],
        name="Max Speed",
    )
```

**Review by rhshadrach — CHANGES_REQUESTED** [review]

**Comment by seanjedi** [comment]
@ramvikrams I think this should resolve the ask that you had: 207abd63cbc36c7fdc2024eaaebc76c1d5336168
Let me know if there is something else i should change

File: pandas/core/groupby/groupby.py [hunk 2]
```diff
@@ -483,7 +490,7 @@ class providing the base-class of operations.
 Examples
 --------
 
->>> df = pd.DataFrame({'A' : ['foo', 'bar', 'foo', 'bar',
+>>> df = pd.%(klass)s({'A' : ['foo', 'bar', 'foo', 'bar',
```
[the remark this answers is not in the document]
Comment (seanjedi) [reply pandas/core/groupby/groupby.py]: I believe this commit should solve this discrepancy: 207abd63cbc36c7fdc2024eaaebc76c1d5336168
I had changed how the template gets made, and made a unique example for both dataframes and series

File: pandas/core/groupby/groupby.py [hunk 3]
```diff
@@ -403,15 +403,22 @@ class providing the base-class of operations.
 f : function, str
     Function to apply to each group. See the Notes section below for requirements.
 
-    Can also accept a Numba JIT function with
-    ``engine='numba'`` specified.
+    Accepted inputs are:
+
+    - string function name
+    - Numba JIT function with ``engine='numba'`` specified.
```
[the remark this answers is not in the document]
Comment (seanjedi) [reply pandas/core/groupby/groupby.py:409]: This should be updated as part of: 207abd63cbc36c7fdc2024eaaebc76c1d5336168
I believe this is now similar to other functions docs we have (IE https://pandas.pydata.org/docs/dev/reference/api/pandas.core.groupby.SeriesGroupBy.aggregate.html)

File: pandas/core/groupby/groupby.py [hunk 4]
```diff
@@ -403,15 +403,22 @@ class providing the base-class of operations.
...
     If the ``'numba'`` engine is chosen, the function must be
     a user defined function with ``values`` and ``index`` as the
     first and second arguments respectively in the function signature.
     Each group's index will be passed to the user defined function
     and optionally available for use.
 
+    If a string is chosen, then it needs to be the string of the
+    function you want to use.
+
+    IE: ``pd.%(klass)s = grouped.transform("sum")``.
```
[the remark this answers is not in the document]
Comment (seanjedi) [reply pandas/core/groupby/groupby.py]: This has been updated in: 207abd63cbc36c7fdc2024eaaebc76c1d5336168

File: pandas/core/groupby/groupby.py [hunk 5]
```diff
@@ -403,15 +403,22 @@ class providing the base-class of operations.
...
+    - Numba JIT function with ``engine='numba'`` specified.
 
     If the ``'numba'`` engine is chosen, the function must be
     a user defined function with ``values`` and ``index`` as the
     first and second arguments respectively in the function signature.
     Each group's index will be passed to the user defined function
     and optionally available for use.
 
+    If a string is chosen, then it needs to be the string of the
+    function you want to use.
```
[the remark this answers is not in the document]
Comment (seanjedi) [reply pandas/core/groupby/groupby.py]: This has been updated in: 207abd63cbc36c7fdc2024eaaebc76c1d5336168

File: pandas/core/groupby/groupby.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rhshadrach) [reply pandas/core/groupby/groupby.py:409]: The list still does not include a standard Python function.

File: pandas/core/groupby/groupby.py [hunk 6]
```diff
@@ -403,15 +403,21 @@ class providing the base-class of operations.
...
 
+    Only passing a single function is supported with this engine.
     If the ``'numba'`` engine is chosen, the function must be
     a user defined function with ``values`` and ``index`` as the
     first and second arguments respectively in the function signature.
     Each group's index will be passed to the user defined function
     and optionally available for use.
 
+    If a string is chosen, then it needs to be the string of the
+    groupby method you want to use.
```
Comment (rhshadrach) [inline pandas/core/groupby/groupby.py:419]: Nit: needs to be the name of the groupby method.

File: pandas/core/groupby/generic.py [hunk 7]
```diff
@@ -429,7 +429,32 @@ def _aggregate_named(self, func, *args, **kwargs):
 
         return result
 
-    @Substitution(klass="Series")
+    __examples_series_doc = dedent(
```
Comment (rhshadrach) [inline pandas/core/groupby/generic.py:432]: The breaking up the examples looks good - can you add an example where the argument is a string (e.g. "mean") to each one.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
Thanks for the update, changes are looking good.

File: pandas/core/groupby/groupby.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (seanjedi) [reply pandas/core/groupby/groupby.py]: Hopefully this is better: 90cae51f589f4a69d41a52308753f81ced25c76d
If not, than can you give me more of an example of what I need to fix? Sorry

File: pandas/core/groupby/groupby.py (hunk 6 above)
Comment (seanjedi) [reply pandas/core/groupby/groupby.py]: Is this better: 90cae51f589f4a69d41a52308753f81ced25c76d

File: pandas/core/groupby/generic.py (hunk 7 above)
Comment (seanjedi) [reply pandas/core/groupby/generic.py:432]: Sure, this was added here: 90cae51f589f4a69d41a52308753f81ced25c76d

**Comment by ramvikrams** [comment]
> @ramvikrams I think this should resolve the ask that you had: 207abd63cbc36c7fdc2024eaaebc76c1d5336168
> Let me know if there is something else i should change



Can you just add the example with string as parameter with its output also

File: pandas/core/groupby/groupby.py [hunk 8]
```diff
@@ -403,15 +403,23 @@ class providing the base-class of operations.
 f : function, str
     Function to apply to each group. See the Notes section below for requirements.
 
-    Can also accept a Numba JIT function with
-    ``engine='numba'`` specified.
+    Accepted inputs are:
+
+    - String function name
+    - Standard Python function
+    - User defined functions IE ``lambda x: x.sum()``
```
Comment (rhshadrach) [inline pandas/core/groupby/groupby.py:410]: These are the same - I think we can just say `Python function`

File: pandas/core/groupby/generic.py [hunk 9]
```diff
@@ -518,6 +548,14 @@ def filter(self, func, dropna: bool = True, *args, **kwargs):
         3    4
         5    6
         Name: B, dtype: int64
+        >>> grouped.transform("mean")
```
Comment (rhshadrach) [inline pandas/core/groupby/generic.py:551]: Looks like this was added to the wrong function; this is the docstring of `filter`.

File: pandas/core/groupby/groupby.py [hunk 10]
```diff
@@ -403,15 +403,23 @@ class providing the base-class of operations.
...
 
+    Only passing a single function is supported with this engine.
     If the ``'numba'`` engine is chosen, the function must be
     a user defined function with ``values`` and ``index`` as the
     first and second arguments respectively in the function signature.
     Each group's index will be passed to the user defined function
     and optionally available for use.
 
+    If a string is chosen, then it needs to be the string name
+    of the groupby method you want to use.
```
Comment (rhshadrach) [inline pandas/core/groupby/groupby.py:421]: I think replacing "string name" with just "name" would be more clear here.

```suggestion
    If a string is chosen, then it needs to be the name
    of the groupby method you want to use.
```

File: pandas/core/groupby/generic.py [hunk 11]
```diff
@@ -429,7 +429,37 @@ def _aggregate_named(self, func, *args, **kwargs):
...
+    Name: Max Speed, dtype: float64
+    >>> grouped.transform("mean")
+    Falcon    370.0
+    Falcon    370.0
+    Parrot     25.0
+    Parrot     25.0
+    """
+    )
+
+    @Substitution(klass="Series", example_class="ser", example=__examples_series_doc)
```
Comment (rhshadrach) [inline pandas/core/groupby/generic.py:462]: Where is example_class used?

**Review by rhshadrach — CHANGES_REQUESTED** [review]

File: pandas/core/groupby/groupby.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (rhshadrach) [reply pandas/core/groupby/groupby.py]: No problem! I'm resolving this conversation, see my request below.

File: pandas/core/groupby/groupby.py (hunk 10 above)
Comment (seanjedi) [reply pandas/core/groupby/groupby.py]: This has been updated here: 9223b59e6aac77dc0a12eb8585ce1e8e0c9f52d4

File: pandas/core/groupby/generic.py (hunk 9 above)
Comment (seanjedi) [reply pandas/core/groupby/generic.py]: Ohh, great catch, I apologize I didn't notice that earlier. This has been resolved here: 9223b59e6aac77dc0a12eb8585ce1e8e0c9f52d4

File: pandas/core/groupby/groupby.py (hunk 8 above)
Comment (seanjedi) [reply pandas/core/groupby/groupby.py]: Alright, hopefully this better: 9223b59e6aac77dc0a12eb8585ce1e8e0c9f52d4

**Comment by seanjedi** [comment]
@rhshadrach Any other changes I should make?

File: pandas/core/groupby/groupby.py [hunk 12]
```diff
@@ -402,15 +402,22 @@ class providing the base-class of operations.
 f : function, str
     Function to apply to each group. See the Notes section below for requirements.
 
-    Can also accept a Numba JIT function with
-    ``engine='numba'`` specified.
+    Accepted inputs are:
+
+    - String function name
```
Comment (rhshadrach) [inline pandas/core/groupby/groupby.py:407]: I think we can just say "String" here; you specify below (L418-420) the requirements of the string.

**Review by rhshadrach — CHANGES_REQUESTED** [review]
This looks really good! One small request

File: pandas/core/groupby/groupby.py (hunk 12 above)
Comment (seanjedi) [reply pandas/core/groupby/groupby.py]: Alright, thanks for the feedback. I have updated the docs here: 04d48f285b9019c71ba77e52170dc69e3c636eee

**Review by rhshadrach — APPROVED** [review]
lgtm, ping on green.

**Review by mroeschke — APPROVED** [review]

**Comment by seanjedi** [comment]
@mroeschke Thanks for merging it, and also sorry for the late reply, I am in grad school and this is finals week so I have been a bit busy today. 
Glad I was able to contribute!

**Pull request merged**